### ISSUE TITLE
Cuda lacks checking of "out of bound"

### ISSUE THREAD

**Issue by ilovepytorch** [pytorch/pytorch#73176]
### 🐛 Describe the bug

Cuda lacks checking of "out of bound" in some APIs. For example,

```python
results = dict()
import torch
arg_class = torch.nn.MultiLabelMarginLoss()
input_tensor = torch.rand([1, 4], dtype=torch.float32)
target_tensor = torch.randint(256,16384,[1, 4], dtype=torch.int64)

try:
  results["res_cpu"] = arg_class(input_tensor.clone(), target_tensor.clone())
except Exception as e:
  results["err_cpu"] = "ERROR:"+str(e)

arg_class = arg_class.cuda()
try:
  results["res_gpu"] = arg_class(input_tensor.clone().cuda(), target_tensor.clone().cuda())
except Exception as e:
  results["err_gpu"] = "ERROR:"+str(e)

print(results)
# {'err_cpu': "ERROR:argument #2 'target' is out of range", 'res_gpu': tensor(6.9519, device='cuda:0')}
```

`MultiLabelMarginLoss` fails in cpu but works in cuda.

Other APIs
```python
torch.nn.functional.cross_entropy
torch.nn.functional.nll_loss
torch.scatter
torch.take
torch.gather
torch.index_select
```

### Versions

PyTorch: 1.10.2
GPU: RTX 3080 Ti
Cuda: 11.3

**Comment by mikeb127** [comment]
I suspect it is because of the differences in error checking in the function multilabel_margin_loss_shape_check between /aten/src/ATen/native/LossMultiLabelMargin.cpp and /aten/src/ATen/native/cuda)/MultiMarginLoss.cu

In particular the following lines in /aten/src/ATen/native/LossMultiLabelMargin.cpp:

`  TORCH_CHECK(
      target.min().item<int64_t>() >= -1, target_arg, " is out of range");
  TORCH_CHECK(
      target.max().item<int64_t>() < dim, target_arg, " is out of range");`

Happy to have a go at this if ok with people.

**Comment by ilovepytorch** [comment]
Any update?

**Comment by mikeb127** [comment]
Seeing as no-one has objected and said they are already working on it, I will make the changes and test them over the next few days. PR be ready to go sometime next week.

**Comment by ngimel** [comment]
@mikeb127 typically we don't do `.min().item()` checks in cuda functions because they cause device synchronization, and instead do kernel asserts where necessary. 
@ilovepytorch please provide minimum reproducible examples for other functions you list.

**Comment by mikeb127** [comment]
@ngimel ok no problems. Will adjust the solution to directly use an assert and avoid .min(),item()

**Comment by mikeb127** [comment]
@ilovepytorch do you have mre's for the other functions available?

PR to fix MultiLabelMarginLoss is ready to go. Happy to add other when available. Otherwise we can merge without them if you like
